pytorch_lightning/trainer/connectors/accelerator_connector.py

```diff
@@ -57,975 +56,801 @@
+                    "You requested one or more GPUs, but set `accelerator='ddp_cpu'`. Training will not use GPUs."
+                )
+        if strategy_flag in ("ddp_spawn", "ddp_spawn_find_unused_parameters_false") and (
+            TorchElasticEnvironment.detect() or KubeflowEnvironment.detect() or self._is_slurm_managing_tasks()
+        ):
+            strategy_flag = "ddp"
+        if strategy_flag in ("dp", "ddp2") and self._accelerator_flag == "cpu":
+            rank_zero_warn(f"{strategy_flag!r} is not supported on CPUs, hence setting `strategy='ddp'`.")
+            strategy_flag = "ddp"
 
-    @property
-    def is_training_type_in_plugins(self) -> bool:
-        return any(
-            (isinstance(plug, str) and plug in StrategyRegistry) or isinstance(plug, Strategy) for plug in self.plugins
-        )
+        if strategy_flag:
+            self._strategy_flag = strategy_flag
 
-    def select_precision_plugin(self) -> PrecisionPlugin:
-        # set precision type
-        self.amp_type = AMPType.from_str(self.amp_type)
+    def _handle_horovod(self) -> None:
+        if self._num_nodes_flag > 1:
+            raise MisconfigurationException(
+                "Horovod does not support setting num_nodes / num_gpus explicitly. Use "
+                "horovodrun / mpirun to configure the number of processes."
+            )
 
-        # validation for all plugins
-        if self.amp_level is not None and self.amp_type != AMPType.APEX:
+        if not _HOROVOD_AVAILABLE:
             raise MisconfigurationException(
-                f"You have asked for `amp_level={self.amp_level!r}` but it's only supported with `amp_backend='apex'`."
+                'Requested `accelerator="horovod"`, but Horovod is not installed.'
+                "Install with \n $HOROVOD_WITH_PYTORCH=1 pip install horovod[pytorch]"
             )
 
-        if self.use_ipu:
-            if self.precision not in (16, 32):
-                raise MisconfigurationException(
-                    f"`Trainer(accelerator='ipu', precision={self.precision!r})` is not supported."
-                )
-            return IPUPrecisionPlugin(self.precision)
-        if self.use_tpu:
-            if self.precision == 32:
+        hvd.init()
+        if isinstance(self.accelerator, GPUAccelerator):
+            # Horovod assigns one local GPU per process
+            self._parallel_devices = list(range(hvd.local_size()))
+        else:
+            self._parallel_devices = [torch.device("cpu")] * hvd.local_size()
+
+    def _init_strategy(self) -> None:
+        """Instantiate the Strategy given depending on the setting of ``_strategy_flag``."""
+        if isinstance(self._strategy_flag, HorovodStrategy) or self._strategy_flag == "horovod":
+            # handle horovod has to happen before initialize strategy because HorovodStrategy needs hvd.init() first.
+            # TODO lazy initialized and setup horovod strategy `global_rank`
+            self._handle_horovod()
```

Comment:
Won't it be cleaner to have `hvd.init()` within HorovodStrategy __init__ method ?

Reply:
We need get device related info to set parallel_devices, which require hvd to be initialized